### DISCUSSION TITLE
What features and enhancements are required for v1.0?

### DISCUSSION THREAD

**Discussion by verseve** [Deltares/Wflow.jl#315] [RFC]
## Crucial features and enhancements for wflow v1.0

- Improve documentation with more examples, including moving to Quarto and a better relationship between code and documentation.
- Better way to deal with the TOML settings, an option could be through the use of Configurations.jl.
- Create documentation for the TOML settings file through a schema and to provide the user with immediate feedback and information while creating the settings file.
- Decouple vertical and horizontal concepts. This also allows for more flexibility and computational efficiency when coupling with external routing or hydrological software packages.
- The `network`, `vertical` and `lateral` fields of the `Model` struct are stored as `NamedTuples`. It would be better to store these variables as `Structs`, this also requires an explicit description of the content. How to best structure this into the `Model` struct also depends on the decoupling of vertical and horizontal concepts and reorganization of long structs (see below).
- Reorganizing of long structs by dividing into smaller sub structs. There is for example quite some overlap between fields of the structs (vertical concepts) `SBM`, `FLEXTOPO` and `HBV`. These are related to the components interception, snow and glacier. While functionality is shared at the low level (vertical_process.jl), this functionality can also be shared at the sub struct level. We could also consider to pull these components out of the vertical struct, e.g. `model.glacier.` etc.
- Reorganizing long functions, this is mainly valid for the `initialize_*_model` functions, that can be separated into multiple sub functions.
- Improve groundwater module. This is now setup from a groundwater model perspective (sort of modflow light), and it would be better to reorganize this. For example it is now not possible to switch between kinematic wave routing (lateral subsurface flow) and Darcian flow (without specifying other groundwater boundaries).
- Lake and reservoir structs and functionality: look into shared functionality and possibly restructure this based on user input. The existing reservoir module is very simple and based on user input may need improvement. 
- Store model or solver settings separately. These settings are now stored as part of structs that also store the variables, for example `kinwave_it` (boolean for the use of iterations as part of the kinematic wave) and `Δt`.
- Support (optional) water balance checks.
- Consistently writing a log file (regardless of how you run a model).
- Unicode characters are used, however we should make a well considered choice whether to use these characters or not (considering advantages and disadvantages). And if we make use of unicode characters it should be consistent throughout the code.
- Different units are used throughout the code (mainly large difference between vertical [mm] and lateral [m] concepts), and unit conversions are hard coded. This can be improved, either by for example sticking to SI units throughout the code or by using a Julia units package as Unitful.jl.
- Internal time step vertical concept: this is now equal to the model time step specified in the TOML file. An improvement could be to use a fixed internal time step (for example hourly), especially considering processes as infiltration and unsaturated flow (for `SBM` iterations are implemented for the unsaturated flow).
- Kinematic wave lateral subsurface flow: this is solved at the specified model time step (TOML file). As for the river and overland flow kinematic wave routing a variable internal time stepping based on the Courant number or a fixed time step specified in the TOML file should be supported.
- The wflow_sbm model is used mostly (internally at Deltares and externally) and most development efforts are aimed at the `SBM` concept. It would be good to decide (and discuss) whether we want to support the `HBV` and `FLEXTOPO` concepts in v1.0.

## Additional features and enhancements for wflow 1.0
These features and enhancements are not critical for v1.0, but would significantly enhance the overall user experience.

- Floodplain routing for the kinematic wave.
- Subgrid functionality for local inertial 1D-2D model.
- Re-infiltration of surface water.
- Adding more refined operational rules for the reservoir module based on global data.
- Enhanced output writing: at specific interval in stead of every time step (maps).

**Comment by markhegnauer** [comment]
Not sure if this discussion is still open, but will give a few extra requests:
- Many NBS type of measures are based on created additional (small scale) storage. It would be nice if wflow could handle this, e.g. by adding a simple threshold level for the water flowing from 1 cell to the other. Such a parameter can then be applied as a map with storage locations / volumes.
-  LAI, as far as I know, is now only linked to interception. Yet, I think LAI also affects evaporation. This would be a nice improvement to the concept, I think.
- I think the reservoir and lakes module requires some refactoring. Typical reservoir functions (e.g. Q-h tables) are actually implemented in the lake module. This is confusing for the user. Above it mentions "based on global data", I think using local data (and making this easy to implement) is even more useful. A connection with Delft-FEWS (where rule curves can be changed based on actual operations) should be made possible. 
- Water use and allocation are very important in many catchments. A good/simple way of implementing this is very high on my wish list. Either through a good-working and simple-to-use coupling to RIBASIM, or (for simple cases) in wflow itself. 
- Output summary: In the section above it mentions output frequency, which I'm very much in favor of. Can this also be expended to a summary at the end of the simulation like it was available in the Python version?

**Comment by markhegnauer** [comment]
One more: the link with water quality! If would be great if a simple tracer could be added to wflow. In the past, there have also been discussions to make it possible to couple DelWaq online. Something similar was done for Delft3D FM, which for water quality was a huge improvement in many cases.
